### ISSUE TITLE
Missing annotations for CDT

### ISSUE THREAD

**Issue by KennethEnevoldsen** [UniversalDependencies/UD_Danish#6]
Comparing the [CDT](https://github.com/mbkromann/copenhagen-dependency-treebank) with DDT, it seems there are annotations available in DDT and CDT which could ideally be combined.

This includes:
- 1) Coreference tags
- 2) QID tags for entities
- 3) Document ids

Notably, the 3) makes a difference while training model for sent. segm. etc. as they are better able to handle long-range dependencies.

A benefit of adding this here would be to have one place to fix all errors at once (see e.g. #4 ). Then datasets such as [DaCoref](https://huggingface.co/datasets/alexandrainst/dacoref) could extract from annotation from DDT to make them more widely available (I have also debated this with @saattrupdan from the Alexandra Institute and sounds like this solution would be meaningful for them as well).

Would love to hear if this is a possibility and if so how I would proceed.

**Comment by dan-zeman** [comment]
If document ids are what they sound to be :-), then it should not be difficult to add them (simple `# newdoc id=XXX` comment before the first sentence of the document) and it would be useful.

Annotation of entities and coreference could be added to the MISC column in the [CorefUD](https://ufal.mff.cuni.cz/corefud) format. CorefUD is a separate project with its own data releases because it includes datasets that cannot be released in UD. But those datasets that also have manual annotation of UPOS tags and dependency relations have the same annotation available also directly in UD, which increases visibility. At present it is the case of English GUM, large part of Czech PDT, Catalan and Spanish AnCora. Some other languages may join the party in the near future. It would be great to have the Danish UD treebank on board, too! (And at the same time DaCoref could join CorefUD.)

**Comment by KennethEnevoldsen** [comment]
CDT does not have doc id's for all the documents in DDT (as it is only a subset), but I assume it does exist somewhere?

Would be great to add DaCoref to CorefUD. As well. From what I understand though annotations would still be separate?

Here is an example of CDT as .conllu.

```
# newdoc id = nw/0442
# sent_id = train-v2-0
# text = På fredag har SID inviteret til reception i SID-huset i anledning af at formanden Kjeld Christensen går ind i de glade tressere.
1	På	på	ADP	_	AdpType=Prep	2	case	_	name=O	id=43	-	nw/0442	-
2	fredag	fredag	NOUN	_	Definite=Ind|Gender=Com|Number=Sing	5	obl	_	name=O	id=44	-	nw/0442	-
3	har	have	AUX	_	Mood=Ind|Tense=Pres|VerbForm=Fin|Voice=Act	5	aux	_	name=O	id=45	-	nw/0442	-
4	SID	SiD	PROPN	_	_	5	nsubj	_	name=B-ORG	id=46	-	nw/0442	Q12335178
5	inviteret	invitere	VERB	_	Definite=Ind|Number=Sing|Tense=Past|VerbForm=Part	0	root	_	name=O	id=47	-	nw/0442	-
6	til	til	ADP	_	AdpType=Prep	7	case	_	name=O	id=48	-	nw/0442	-
7	reception	reception	NOUN	_	Definite=Ind|Gender=Com|Number=Sing	5	obl	_	name=O	id=49	-	nw/0442	-
8	i	i	ADP	_	AdpType=Prep	9	case	_	name=O	id=50	-	nw/0442	-
9	SID-huset	SID-hus	NOUN	_	Definite=Def|Gender=Neut|Number=Sing	7	nmod	_	name=B-LOC	```

**Comment by dan-zeman** [comment]
I suppose it would yield something like the following in both CorefUD and UD. A valid CoNLL-U file, so it could be directly THE UD treebank, not a separate annotation file. I don't know what "Q12335178" means in the CDT example but if it is an identification of entity coreference cluster, then it would be reflected in the entity id (e1 would be used again when any reference to SID is made).

```conllu
# global.Entity = eid-etype-head-other
# newdoc id = nw/0442
# sent_id = train-v2-0
# text = På fredag har SID inviteret til reception i SID-huset i anledning af at formanden Kjeld Christensen går ind i de glade tressere.
1	På	på	ADP	_	AdpType=Prep	2	case	_	_
2	fredag	fredag	NOUN	_	Definite=Ind|Gender=Com|Number=Sing	5	obl	_	_
3	har	have	AUX	_	Mood=Ind|Tense=Pres|VerbForm=Fin|Voice=Act	5	aux	_	_
4	SID	SiD	PROPN	_	_	5	nsubj	_	Entity=(e1-organization-1)
5	inviteret	invitere	VERB	_	Definite=Ind|Number=Sing|Tense=Past|VerbForm=Part	0	root	_	_
6	til	til	ADP	_	AdpType=Prep	7	case	_	_
7	reception	reception	NOUN	_	Definite=Ind|Gender=Com|Number=Sing	5	obl	_	_
8	i	i	ADP	_	AdpType=Prep	9	case	_	_
9	SID-huset	SID-hus	NOUN	_	Definite=Def|Gender=Neut|Number=Sing	7	nmod	_	Entity=(e2-place-1)

```

As for CDT being only a subset of DDT: I happen to have a copy of something called ddt-1.0.zip on my disk :-) and it seems to be in TEI-XML format with (probably) more than enough information to restore the documents. I don't have capacity to work on it but I'll be happy to share the files if you want to give it a try.

**Comment by KennethEnevoldsen** [comment]
QID is the corespondance to the wikidata entry (though I don't entirely agree with its current annotation scheme). So no not a coref ID.

here is an example with coreferences, e.g. `SF`, `SF og Fremskridtspartiet`, `Fremskridtspartiet`:
```python
53	både	både	ADV	_	_	54	advmod	_	name=O	id=253	-	nw/0442	-
54	SF	SF	PROPN	_	_	57	nsubj	_	name=B-ORG	id=254	(197)|(1845	nw/0442	Q615603
55	og	og	CCONJ	_	_	56	cc	_	name=O	id=255	-	nw/0442	-
56	Fremskridtspartiet	fremskridtsparti	NOUN	_	Definite=Def|Gender=Neut|Number=Sing	54	conj	_	name=B-ORG	id=256	(197)|1845)	nw/0442	Q1455237
57	fået	få	VERB	_	Definite=Ind|Number=Sing|Tense=Past|VerbForm=Part	38	conj	_	name=O	id=257	-	nw/0442	-
58	deres	deres	DET	_	Number[psor]=Plur|Person=3|Poss=Yes|PronType=Prs	59	det	_	name=O	id=258	(1845)|(197)	nw/0442	-
```
Note that they have the same coref ID but not the same QID.

Please share the files. Would love to take a look at it. Can try to see how much it influence the downstream model performance as well

edit: You can contact me using kennethenevoldsen@gmail.com

**Comment by dan-zeman** [comment]
I think that the second example would look like this in the CorefUD format:

```conllu
53	både	både	ADV	_	_	54	advmod	_	_
54	SF	SF	PROPN	_	_	57	nsubj	_	Entity=(e5-organization-1(e3-organization-1-Q615603)
55	og	og	CCONJ	_	_	56	cc	_	_
56	Fremskridtspartiet	fremskridtsparti	NOUN	_	Definite=Def|Gender=Neut|Number=Sing	54	conj	_	Entity=(e4-organization-1-Q1455237)e5)
57	fået	få	VERB	_	Definite=Ind|Number=Sing|Tense=Past|VerbForm=Part	38	conj	_	_
58	deres	deres	DET	_	Number[psor]=Plur|Person=3|Poss=Yes|PronType=Prs	59	det	_	Entity=(e5-organization-1)

```

> QID is the corespondance to the wikidata entry

Nice. In UD, such anchoring currently exists in [UD_English-GUM](https://raw.githubusercontent.com/UniversalDependencies/UD_English-GUM/master/en_gum-ud-train.conllu) but instead of Wikidata, they refer to the name of the article in the English Wikipedia. I think that Wikidata is better because it is more portable across languages, even if the identifier is not self-explanatory.

> Note that they have the same coref ID but not the same QID.

This does not make sense to me. If two entities have different identity, their mentions cannot be coreferential. I agree with the coordination being marked as a mention of entity 1845, and _deres_ being coreferential with the whole coordination, hence having 1845 as well. It is also understandable that _SF_ and _Fremskridtspartiet_ have two different Wikidata ids – they are two different parties. But then their coreference id should differ as well, e.g., one should be 197 and the other 198. And _deres_ should be neither 197 nor 198, provided that it is coreferential with both of them and not just one.

> Please share the files.

Link sent via e-mail.

**Comment by KennethEnevoldsen** [comment]
> deres should be neither 197 nor 198, provided that it is coreferential with both of them and not just one.

I tend to agree with your point, though I did not tag the data, but I think that the idea is that 'deres' refers to both of them and each of them independently. Tough I am not sure I agree with the last annotation though. Since they are never referred to independently there should not be a coref annotation on them independently (Again, I am not sure whether I agree with this).

**Comment by KennethEnevoldsen** [comment]
Oh and seems like I misspelled my email. Could you resend it to:
[kennethcenevoldsen@gmail.com](mailto:kennethcenevoldsen@gmail.com)

**Comment by dan-zeman** [comment]
> Oh and seems like I misspelled my email. Could you resend it to: [kennethcenevoldsen@gmail.com](mailto:kennethcenevoldsen@gmail.com)

Resent :-)

**Comment by amir-zeldes** [comment]
> Nice. In UD, such anchoring currently exists in [UD_English-GUM](https://raw.githubusercontent.com/UniversalDependencies/UD_English-GUM/master/en_gum-ud-train.conllu) but instead of Wikidata, they refer to the name of the article in the English Wikipedia. I think that Wikidata is better because it is more portable across languages, even if the identifier is not self-explanatory.

Hi, just chiming in to say we actually also have Wikidata IDs for all entity linking annotations, the mapping is kept up to date with Wikipedia article title changes if relevant and can be found here:

https://github.com/amir-zeldes/gum/blob/master/coref/wiki_map.tab

The DateUpdated column gives the last check time at which the Wikidata identifier alignment was verified.

> If two entities have different identity, their mentions cannot be coreferential

This is also the way we do things for GUM, and it is enforced by a validator. That said, I can imagine a conceivable use case that deviates from that. For example, Wikipedia has a separate article for monotheistic [God](https://en.wikipedia.org/wiki/God), and a separate one for [Lord](https://en.wikipedia.org/wiki/Lord), which includes a section on the use of that appellation to refer to God. We once had a document with both forms referring to God, which were annotated as coref. But then the Wikification annotators linked the mentions to different identifiers (God vs. Lord), which the validator caught. After discussion, we decided to make both point to "God", which seemed like a better match, but I could imagine a corpus where cases like this or similar might link to near synonym entries based on the text form of the mention (again, this is categorically prohibited in the GUM family of corpora, I am just saying it's conceivable).

> deres should be neither 197 nor 198, provided that it is coreferential with both of them and not just one.

I definitely agree with this one, coordination creates a new mention, and if there is a split antecedent, that should require a different type of annotation.

**Comment by KennethEnevoldsen** [comment]
Thanks for this @amir-zeldes. It would be nice to standardise the Danish format to this as well, I know that we have some more coreference data coming out from our lab. Is there a specific annotation guideline that you follow?

**Comment by amir-zeldes** [comment]
Yes - the GUM-specific coreference guidelines are [here](https://wiki.gucorpling.org/gum/entities) and are harmonized with the entity annotation guidelines in terms of markable selection etc. They are also very similar to the guidelines used by the ARRAU corpus for English. 

The guidelines cover a variety of phenomena not covered in other coref datasets, which usually annotate only a subset of these. Probably the most popular and widely implemented guidelines are those used by OntoNotes, which are a subset of what GUM covers, and in fact we have an automatically converted version of the data reduced to include only those phenomena covered by OntoNotes, for compatibility (more details in [this paper](https://aclanthology.org/2021.crac-1.15/))

Last year I also wrote a position paper comparing annotation schemes and pointing out the limits of the OntoNotes scheme. It goes over some worked examples which could be helpful in deciding what to include, and you can find it here:

  * [Can we Fix the Scope for Coreference? Problems and Solutions for Benchmarks beyond OntoNotes](https://journals.uic.edu/ojs/index.php/dad/article/view/11706/10635)

Hope that helps!

**Comment by KennethEnevoldsen** [comment]
Wonderful! Exactly what I wanted to see
